plugin/CactbotEventSource/FateWatcher.cs

```diff
@@ -138,68 +183,131 @@ private struct OPCodes {
       throw new Exception($"Enum value {name} not found in {type}!");
     }
 
-    private ushort GetOpcode(string name) {
+    private static ushort GetOpcode(Type type, string name) {
       // FFXIV_ACT_Plugin 2.0.4.14 converted Server_MessageType from enum to struct. Deal with each type appropriately.
-      if (MessageType.IsEnum) {
-        return (ushort)GetEnumValue(MessageType, name);
+      if (type.IsEnum) {
+        return (ushort)GetEnumValue(type, name);
       } else {
-        var value = MessageType.GetField(name).GetValue(null);
+        var value = type.GetField(name).GetValue(null);
         return (ushort)value.GetType().GetProperty("InternalValue").GetValue(value);
       }
     }
 
     private unsafe void MessageReceived(string id, long epoch, byte[] message) {
-      if (message.Length < ActorControl143_Size)
+      if (message.Length < actorControl143.size && message.Length < cedirectoropcodes[region_].size)
         return;
 
       fixed (byte* buffer = message) {
-        if (*((ushort*)&buffer[MessageType_Offset]) == ActorControl143_Opcode) {
-          ProcessMessage(buffer, message);
+        if (*(ushort*)&buffer[messageTypeOffset] == actorControl143.opCode) {
+          ProcessActorControl143(buffer, message);
+          return;
+        }
+        if (cedirectoropcodes.ContainsKey(region_)) {
+          if (*(ushort*)&buffer[messageTypeOffset] == cedirectoropcodes[region_].opcode) {
+            ProcessCEDirector(buffer, message);
+            return;
+          }
         }
       }
     }
 
-    public unsafe void ProcessMessage(byte* buffer, byte[] message) {
-      int a = *(ushort*)&buffer[Category_Offset];
-
-      if (a == opcodes[region_].add) {
-        AddFate(*(int*)&buffer[Param1_Offset]);
-      } else if (a == opcodes[region_].remove) {
-        RemoveFate(*(int*)&buffer[Param1_Offset]);
-      } else if (a == opcodes[region_].update) {
-        int param1 = *(int*)&buffer[Param1_Offset];
-        int param2 = *(int*)&buffer[Param2_Offset];
-        if (!fates.ContainsKey(param1)) {
-          AddFate(param1);
+    public unsafe void ProcessActorControl143(byte* buffer, byte[] message) {
+      int a = *(ushort*)&buffer[actorControl143.categoryOffset];
+
+      fateSemaphore.WaitAsync();
```

Comment:
Each network packet is handled in it's own thread, and from testing I was able to periodically get into a state where stale data was overwriting new data into the dictionaries. It would be possible to queue these up and process them in a single thread, but each transaction is very small (the longest probably being ```!ces[data.ceKey].Equals(data)```). I'm happy to go with your decision here but I think the mutexes are enough for now.